# Bi and Tri Grams

In [1]:
import pandas as pd
#df_trial_essays = pd.read_csv(r"C:\Users\Admin\HC3 EDA\prepped essays.csv") # not good enoguh

>>> fit for bigram eval bec in order of sentences. Annoying because removed 'pros' and 'cons' . May be a better dataset. 
FOr phrase eval want to have some nouns , althouhg not proper nouns . So really need the step befroe 'prepped' 

In [2]:
import pandas as pd

In [3]:
# reddit clean
df_reddit_clean = pd.read_csv(r"C:\Users\Admin\HC3 EDA\GriD\reddit_bigram_ready.csv")

In [4]:
df_reddit_clean

word
0            The
1       standoff
2             in
3            was
4              a
...          ...
100205     least
100206       the
100207   regular
100208   minimum
100209      wage

[100210 rows x 1 columns]

In [5]:
# answers
answer1 = pd.read_csv(r"C:\Users\Admin\HC3 EDA\answers_clean1.csv")
answer1 = answer1.drop(columns=['Unnamed: 0'])


In [6]:
# essays 
essay1 = pd.read_csv(r"C:\Users\Admin\HC3 EDA\essays_clean1.csv")
essay1 =essay1.drop(columns =['Unnamed: 0'])

In [7]:
answer1 = answer1.dropna()


In [8]:
essay1 = essay1.dropna()


In [9]:
answer1 = answer1.rename(columns={'col1': 'word'})
answer1

word
0              when
1       considering
2               the
3              pros
4               and
...             ...
538476  environment
538477          for
538478        teens
538479          and
538480     families

[486524 rows x 1 columns]

In [10]:
essay1 =essay1.rename(columns={'col1' :'word'})
essay1

word
0           there
1             are
2            many
3       different
4            best
...           ...
20892  understand
20893         the
20894       world
20895      around
20896          us

[18724 rows x 1 columns]

In [11]:
# concat all into one big df 
# first test:
all_3 = pd.concat([df_reddit_clean]+[essay1]+[answer1])

In [12]:
all_3

word
0               The
1          standoff
2                in
3               was
4                 a
...             ...
538476  environment
538477          for
538478        teens
538479          and
538480     families

[605458 rows x 1 columns]

In [13]:
# save to csv just in case
all_3.to_csv('ngram_prepped_merged.csv', index=False)

> Using all_3 df for bigram eval . 

>>> skipping paraphrases

# PLAN : KEEP ALL NOUNS and manually remove after tri and bigram eval , so not to lose meaningful , general phrases w nouns . Remove proper nouns.
# MAYBE KEEP THE FUNCTIONAL WORDS ( like sapling.ai)


# Bi-grams 

In [14]:
# trying one way
def find_ngrams(input_list, n):
    return list(zip(*[input_list[i:] for i in range(n)]))


all_3['bigrams'] = all_3['word'].map(lambda x: find_ngrams(x, 2))
all_3.head()




word                                            bigrams
0       The                                   [(T, h), (h, e)]
1  standoff  [(s, t), (t, a), (a, n), (n, d), (d, o), (o, f...
2        in                                           [(i, n)]
3       was                                   [(w, a), (a, s)]
4         a                                                 []

In [15]:
import pandas as pd
all_3 = pd.read_csv(r"C:\Users\Admin\HC3 EDA\GriD\ngram_prepped_merged.csv")

# for wor din all3  " ".join

In [16]:
all_3

word
0               The
1          standoff
2                in
3               was
4                 a
...             ...
605453  environment
605454          for
605455        teens
605456          and
605457     families

[605458 rows x 1 columns]

In [17]:
one_string = ' '.join(all_3['word'].tolist())
one_string1 =  one_string.lower() # dont need for loop becasue .lower() lowercases the entire string 
# one_string1

In [18]:
# save to text file
# text_file = open(r'one string.txt', 'w')

# text_file.write(one_string1)
# text_file.close()


In [19]:
# now trying bigrams with method # 1

# A : Make df of string

all_3['bigrams'] = all_3['word'].map(lambda x: find_ngrams(x, 2))
all_3.head()


word                                            bigrams
0       The                                   [(T, h), (h, e)]
1  standoff  [(s, t), (t, a), (a, n), (n, d), (d, o), (o, f...
2        in                                           [(i, n)]
3       was                                   [(w, a), (a, s)]
4         a                                                 []

In [20]:
ng =find_ngrams(one_string1,2)
ng

[('t', 'h'),
 ('h', 'e'),
 ('e', ' '),
 (' ', 's'),
 ('s', 't'),
 ('t', 'a'),
 ('a', 'n'),
 ('n', 'd'),
 ('d', 'o'),
 ('o', 'f'),
 ('f', 'f'),
 ('f', ' '),
 (' ', 'i'),
 ('i', 'n'),
 ('n', ' '),
 (' ', 'w'),
 ('w', 'a'),
 ('a', 's'),
 ('s', ' '),
 (' ', 'a'),
 ('a', ' '),
 (' ', 'c'),
 ('c', 'o'),
 ('o', 'm'),
 ('m', 'p'),
 ('p', 'l'),
 ('l', 'e'),
 ('e', 'x'),
 ('x', ' '),
 (' ', 'a'),
 ('a', 'n'),
 ('n', 'd'),
 ('d', ' '),
 (' ', 'c'),
 ('c', 'o'),
 ('o', 'n'),
 ('n', 't'),
 ('t', 'r'),
 ('r', 'o'),
 ('o', 'v'),
 ('v', 'e'),
 ('e', 'r'),
 ('r', 's'),
 ('s', 'i'),
 ('i', 'a'),
 ('a', 'l'),
 ('l', ' '),
 (' ', 'e'),
 ('e', 'v'),
 ('v', 'e'),
 ('e', 'n'),
 ('n', 't'),
 ('t', ' '),
 (' ', 't'),
 ('t', 'h'),
 ('h', 'a'),
 ('a', 't'),
 ('t', ' '),
 (' ', 'g'),
 ('g', 'a'),
 ('a', 'r'),
 ('r', 'n'),
 ('n', 'e'),
 ('e', 'r'),
 ('r', 'e'),
 ('e', 'd'),
 ('d', ' '),
 (' ', 's'),
 ('s', 'i'),
 ('i', 'g'),
 ('g', 'n'),
 ('n', 'i'),
 ('i', 'f'),
 ('f', 'i'),
 ('i', 'c'),
 ('c', 'a'),
 ('a', 'n'),

In [21]:
# method #2
import nltk
from nltk import ngrams
sentence = input(one_string1)
n = int(input("2"))
n_grams = ngrams(sentence.split(), n)
for grams in n_grams:
    print(grams)

In [ ]:
from nltk.collocations import BigramCollocationFinder
from nltk.metrics import BigramAssocMeasures

In [ ]:
# Create a BigramCollocationFinder
finder = BigramCollocationFinder.from_words(one_string1)
found = finder.nbest(bigram_measures.pmi, 100)
found

[('₁', '₀'),
 ('6', '°'),
 ('0', '^'),
 ('2', '0'),
 ('6', '4'),
 ('8', '6'),
 ('3', '6'),
 ('4', '4'),
 ('3', '3'),
 ('3', '.'),
 ('.', '7'),
 ('3', '0'),
 ('6', '0'),
 ('.', '6'),
 (':', '1'),
 ('7', '4'),
 ('8', '7'),
 ('9', '6'),
 ('7', '0'),
 ('1', '8'),
 ('1', '9'),
 ('9', '4'),
 ('9', '3'),
 ('7', '9'),
 ('9', '7'),
 ('7', '7'),
 ('^', '-'),
 ('2', '3'),
 ('3', '5'),
 ('6', '7'),
 ('2', '.'),
 ('9', '2'),
 ('6', '8'),
 ('2', '4'),
 ('1', '7'),
 ('4', '0'),
 ('4', '9'),
 ('1', '6'),
 ('2', '2'),
 ('8', '5'),
 ('9', '9'),
 ('5', '0'),
 ('2', '7'),
 ('1', '4'),
 ('3', '9'),
 ('0', '0'),
 ('4', '-'),
 ('1', '3'),
 ('4', '5'),
 ('0', '6'),
 ('6', '9'),
 ('8', '0'),
 ('2', '9'),
 ('8', '9'),
 ('9', '8'),
 ('-', '3'),
 ('9', '1'),
 ('7', '-'),
 ('6', '5'),
 ('-', '2'),
 ('5', '8'),
 ('0', '-'),
 ('8', '-'),
 ('9', '5'),
 ('1', '0'),
 ('1', '2'),
 ('-', '1'),
 ('-', '6'),
 ('6', '-'),
 ("'", "'"),
 ('2', '-'),
 ('6', '1'),
 ('9', '-'),
 ('2', '1'),
 ('1', '1'),
 ('-', '5'),
 ('1', '5'),

In [ ]:
finder

# need to prep more. remove punct 

In [ ]:

# Get the most common bigrams
bigram_measures = nltk.collocations.BigramAssocMeasures()
found_bigrams = finder.nbest(bigram_measures.pmi, 20)

In [ ]:
found_bigrams

In [ ]:
# gives phrases but no rank 
# from nltk.util import ngrams

# n = 2
# sentence = one_string1
# unigrams = ngrams(sentence.split(), n)

# for item in unigrams:
#     print(item)

## ^ works but does not give frequency or pmi .

In [ ]:
# length of one_string1
len(one_string1)

3622049

# ^ for article

In [ ]:
# method #3 
import nltk
bigrams = nltk.collocations.BigramAssocMeasures()
#trigrams = nltk.collocations.TrigramAssocMeasures()
bigramFinder = nltk.collocations.BigramCollocationFinder.from_words(one_string1.split())
#trigramFinder = nltk.collocations.TrigramCollocationFinder.from_words(one_string1)

In [ ]:
# cont'
#bigrams
bigram_freq = bigramFinder.ngram_fd.items()
bigramFreqTable = pd.DataFrame(list(bigram_freq), columns=['bigram','freq']).sort_values(by='freq', ascending=False)
#trigrams
#trigram_freq = trigramFinder.ngram_fd.items()
#trigramFreqTable = pd.DataFrame(list(trigram_freq), columns=['trigram','freq']).sort_values(by='freq', ascending=False)

In [ ]:
bigramFreqTable

bigram  freq
21                  (it, is)  2400
30                 (of, the)  2285
1179               (can, be)  2258
74                 (in, the)  2215
970                (on, the)  1456
...                      ...   ...
73361        (rebellion, in)     1
73360        (of, rebellion)     1
73357          (trusted, or)     1
73356         (not, trusted)     1
175161  (circumstances, but)     1

[175162 rows x 2 columns]

In [ ]:
bigramFreqTable.head(20)

bigram  freq
21           (it, is)  2400
30          (of, the)  2285
1179        (can, be)  2258
74          (in, the)  2215
970         (on, the)  1456
565        (such, as)  1212
51          (to, the)  1155
964           (in, a)  1062
23    (important, to)  1019
983          (to, be)  1010

In [ ]:
# v2 . Try to find pmi and freq 
# finder.apply_freq_filter(3)
# finder.nbest(bigram_measures.pmi, 10)
import nltk
from nltk.collocations import *
bigram_measures = nltk.collocations.BigramAssocMeasures()
trigram_measures = nltk.collocations.TrigramAssocMeasures()

# change this to read in your data
finder = BigramCollocationFinder.from_words(one_string1.split())
finder2 = TrigramCollocationFinder.from_words(one_string1.split())

# only bigrams that appear 10+ times
finder.apply_freq_filter(10)
# only trigrams that appear  10+ time
finder2.apply_freq_filter(10)

# quadgram
quadgram_measures = nltk.collocations.QuadgramAssocMeasures()
finder3 =QuadgramCollocationFinder.from_words(one_string1.split())
finder3.apply_freq_filter(10)
# return the 10 n-grams with the highest PMI
#print(finder.nbest(bigram_measures.pmi, 1000))
# print

In [ ]:
# from collections import _itemgetter

In [ ]:
# from https://stackoverflow.com/questions/25924932/find-frequency-and-pmi-score-of-a-bigram-using-nltk-python-3?rq=3  
# maybe play around with and modifiy functions to include freq

# def nbest(self, score_fn, n):
#     """Returns the top n ngrams when scored by the given function."""
#     return [p for p,s in self.score_ngrams(score_fn)[:n]]
# #So you could call the score_ngrams() directly without getting the nbest since it returns a sorted list anyways.:

# def score_ngrams(self, score_fn1 , score_fn2):
#     """Returns a sequence of (ngram, score) pairs ordered from highest to
#     lowest score, as determined by the scoring function provided.

#     """
#     # freq = FredDist(self )
#     return sorted(self._score_ngrams(score_fn1),self._score_ngrams(score_fn2),
#                   key=_itemgetter(1), reverse=True)

In [ ]:
# # input raw_freq
# for i in finder.score_ngrams(bigram_measures.raw_freq):
#     print(i)
# print()

In [ ]:
# # gives pmi for bigrams 
# # based on above function

# for i in finder.score_ngrams(bigram_measures.pmi):
#     print(i)
# # print()

In [ ]:
bigram_freq1 = finder.ngram_fd.items()
bigram_freq1

In [ ]:
trigram_freq1 = finder2.ngram_fd.items()
trigram_freq1

In [ ]:
quadgram_freq1 = finder3.ngram_fd.items()
quadgram_freq1

In [ ]:
for item in bigram_freq1:
    bigram, pmi = item
    print(bigram)

  
        

In [ ]:
# bigram_freq = bigramFinder.ngram_fd.items()
bigram_pmi = [i for i in finder.score_ngrams(bigram_measures.pmi)] # [(( phrase1, phrase2 ), pmi) ]
#bigramFreqTable = pd.DataFrame(list(bigram_freq1), columns=['bigram','freq']).sort_values(by='freq', ascending=False)

In [ ]:
trigram_pmi = [i for i in finder2.score_ngrams(trigram_measures.pmi)] # [(( phrase1, phrase2 ), pmi) ]

In [ ]:
quadgram_pmi = [i for i in finder3.score_ngrams(quadgram_measures.pmi)]

In [ ]:
quadgram_pmi

In [ ]:
trigram_pmi

In [ ]:
bigram_freq1

In [ ]:
# bigram list creation
pmi_list =[]
count_list =  []
phrase_list =[]
bigram_list = [] # to check if bigram order = phrase list order
for phrase , pmi in bigram_pmi:
    for item in bigram_freq1:
        bigram, count = item
        if bigram == phrase:
            pmi_list.append(pmi) # pmi for the bigram  = phrase
            count_list.append(count)  # count for the phrase = bigram 
            phrase_list.append(phrase) # phrase = bigram
            bigram_list.append(bigram)
            

In [ ]:
# trigram list creation

pmi2_list =[]
count2_list =  []
phrase2_list =[]
bigram2_list = [] # to check if bigram order = phrase list order
for phrase , pmi in trigram_pmi:
    for item in trigram_freq1:
        bigram, count = item
        if bigram == phrase:
            pmi2_list.append(pmi) # pmi for the bigram  = phrase
            count2_list.append(count)  # count for the phrase = bigram 
            phrase2_list.append(phrase) # phrase = bigram
            bigram2_list.append(bigram)
            

In [ ]:
# quadgram list creation
pmi3_list =[]
count3_list =[]
phrase3_list =[]
for phrase, pmi in quadgram_pmi:
    for item in quadgram_freq1:
        quadgram,count = item
        if quadgram == phrase:
            pmi3_list.append(pmi)
            count3_list.append(count)
            phrase3_list.append(phrase)

In [ ]:
#checking
for item in bigram_freq1:
        bigram, count = item
        if bigram == ('fossil', 'fuels'):
            print(item)

In [ ]:
bigram_list

In [ ]:
#checking
bigram_pmi

In [ ]:
#df_pcpmi_bigram = pd.DataFrame(columns=['Bi-Phrase','Count','PMI']).sort_values(by='PMI', ascending=False)
df_bigram = pd.DataFrame({'Phrase':phrase_list , 'Count':count_list , 'PMI':pmi_list}) # how to sort this by pmi

In [ ]:
df_trigram = pd.DataFrame({'Phrase':phrase2_list , 'Count' :count2_list, 'PMI':pmi2_list})

In [ ]:
df_quadgram = pd.DataFrame({'Phrase': phrase3_list, 'Count': count3_list, 'PMI':pmi3_list})

In [ ]:
df_quadgram = df_quadgram.sort_values(by='Count', ascending=False)

In [ ]:
df_quadgram

Phrase  Count        PMI
1336           (the, students, and, the)     13   5.476013
1335         (and, the, opportunity, to)     12   8.164491
1334                 (and, out, of, the)     11   8.614202
1333                  (and, it, can, be)     19   8.804104
1332       (and, understanding, of, the)     12   8.886959
...                                  ...    ...        ...
4        (after, school, homework, club)     34  27.850734
3          (and, early, 20th, centuries)     11  27.860815
2               (setting, our, aim, too)     19  27.918236
1     (seeking, guidance, from, experts)     18  28.166498
0                   (our, aim, too, low)     10  29.144632

[1337 rows x 3 columns]

In [ ]:
df_quadgram.to_csv('quadgram1.csv', index=False)

In [ ]:
# remove parentheses 
clean_phrase = []
for phrases in df_quadgram['Phrase']:
    no_brack = ' '.join(phrases)
    clean_phrase.append(no_brack)
# my_tuple = ('Python', 'Tuple', 'To', 'String')
# result_string = ' '.join(my_tuple)
# print(result_string)

In [ ]:
df_quadgram['Phrase'] = clean_phrase

In [ ]:
df_quadgram.to_csv('quadgram1.csv', index=False)

In [ ]:
df_trigram = df_trigram.sort_values(by='Count', ascending=False)


In [ ]:
df_quadgram

Phrase  Count        PMI
1335         and the opportunity to     12   8.164491
1334                 and out of the     11   8.614202
1333                  and it can be     19   8.804104
1332       and understanding of the     12   8.886959
1331               of the world and     15   8.914403
...                             ...    ...        ...
4        after school homework club     34  27.850734
3          and early 20th centuries     11  27.860815
2               setting our aim too     19  27.918236
1     seeking guidance from experts     18  28.166498
0                   our aim too low     10  29.144632

[1336 rows x 3 columns]

In [ ]:
df_trigram

Count        PMI                Phrases
0       788  11.326107      it  is  important
1       694   9.355931      is  important  to
2       512  10.086778         on  the  other
3       512  10.131633          in  order  to
4       496  12.661344       the  other  hand
...     ...        ...                    ...
4323     10   8.257245    attitude  can  make
4324     10   8.249711         it  harder  to
4325     10   8.228961         will  need  to
4326     10   8.196099  both  beneficial  and
4327     10  -2.093969           the  to  the

[4328 rows x 3 columns]

In [ ]:
df_quadgram

Phrase  Count        PMI
1336           the students and the     13   5.476013
1335         and the opportunity to     12   8.164491
1334                 and out of the     11   8.614202
1333                  and it can be     19   8.804104
1332       and understanding of the     12   8.886959
...                             ...    ...        ...
4        after school homework club     34  27.850734
3          and early 20th centuries     11  27.860815
2               setting our aim too     19  27.918236
1     seeking guidance from experts     18  28.166498
0                   our aim too low     10  29.144632

[1337 rows x 3 columns]

In [ ]:
df_quadgram.drop(df_quadgram.loc[df_quadgram['Phrase']=="the students and the"].index, inplace=True)

In [ ]:
df_bigram = df_bigram.sort_values(by='Count', ascending=False)


In [ ]:
df_bigram.head(50)

In [ ]:
df_bigram.to_csv('bigram phrases.csv')

In [ ]:
df_trigram.to_csv('trigram phrases.csv')

In [ ]:
import pandas as pd
df_trigram = pd.read_csv(r'C:\Users\Admin\HC3 EDA\GriD\trigram phrases.csv')

In [ ]:
# unnecessary
# gives count

# create a bigram FreqDist object
# bigrams = nltk.bigrams(one_string1.split())
# freq_dist = FreqDist(bigrams) 
# # item
# print('Unique elements in corpus and their counts:')
# for i in freq_dist.items():
#   print(i)
# print()

In [ ]:
# combine the phrase parts into one string
df_trigram

Unnamed: 0                         Phrase  Count        PMI
0           1120      ('it', 'is', 'important')    788  11.326107
1           2092      ('is', 'important', 'to')    694   9.355931
2           1703         ('on', 'the', 'other')    512  10.086778
3           1673          ('in', 'order', 'to')    512  10.131633
4            658       ('the', 'other', 'hand')    496  12.661344
...          ...                            ...    ...        ...
4323        2695    ('attitude', 'can', 'make')     10   8.257245
4324        2699         ('it', 'harder', 'to')     10   8.249711
4325        2715         ('will', 'need', 'to')     10   8.228961
4326        2731  ('both', 'beneficial', 'and')     10   8.196099
4327        4327           ('the', 'to', 'the')     10  -2.093969

[4328 rows x 4 columns]

In [ ]:
neat_phrases = []

for tup in df_trigram['Phrase']:
    phrase = " ".join(tup[1:-1].split(",")) 
    phrase = phrase.replace("'", "")
    neat_phrases.append(phrase)



In [ ]:
neat_phrases

In [ ]:
df_trigram['Phrases'] = neat_phrases

In [ ]:
df_trigram

Unnamed: 0                         Phrase  Count        PMI  \
0           1120      ('it', 'is', 'important')    788  11.326107   
1           2092      ('is', 'important', 'to')    694   9.355931   
2           1703         ('on', 'the', 'other')    512  10.086778   
3           1673          ('in', 'order', 'to')    512  10.131633   
4            658       ('the', 'other', 'hand')    496  12.661344   
...          ...                            ...    ...        ...   
4323        2695    ('attitude', 'can', 'make')     10   8.257245   
4324        2699         ('it', 'harder', 'to')     10   8.249711   
4325        2715         ('will', 'need', 'to')     10   8.228961   
4326        2731  ('both', 'beneficial', 'and')     10   8.196099   
4327        4327           ('the', 'to', 'the')     10  -2.093969   

                    Phrases  
0         it  is  important  
1         is  important  to  
2            on  the  other  
3             in  order  to  
4          the  other  hand  
...                     ...  
4323    attitude  can  make  
4324         it  harder  to  
4325         will  need  to  
4326  both  beneficial  and  
4327           the  to  the  

[4328 rows x 5 columns]

In [ ]:
df_trigram.drop(columns=['Unnamed: 0' , 'Phrase'], inplace=True)

In [ ]:
df_trigram

Count        PMI                Phrases
0       788  11.326107      it  is  important
1       694   9.355931      is  important  to
2       512  10.086778         on  the  other
3       512  10.131633          in  order  to
4       496  12.661344       the  other  hand
...     ...        ...                    ...
4323     10   8.257245    attitude  can  make
4324     10   8.249711         it  harder  to
4325     10   8.228961         will  need  to
4326     10   8.196099  both  beneficial  and
4327     10  -2.093969           the  to  the

[4328 rows x 3 columns]

In [ ]:
df_trigram.to_csv('ready trigrams.csv',index=False)

In [ ]:
# look for low pmi and high frequency COMMON CHATGPT UNCOMMON HUMAN
low_pmi_df = df_trigram.loc[df_trigram['PMI'] < 5] 

In [ ]:
# now for quad
low_pmi_df2 = df_quadgram.loc[df_quadgram['PMI'] < 10] # for quad, the lowest pmi score is above 8 !

In [ ]:
df_quadgram= df_quadgram.sort_values("PMI")

In [ ]:
df_quadgram.head(20)

Phrase  Count        PMI
1336      (the, students, and, the)     13   5.476013
1335    (and, the, opportunity, to)     12   8.164491
1334            (and, out, of, the)     11   8.614202
1333             (and, it, can, be)     19   8.804104
1332  (and, understanding, of, the)     12   8.886959
1331          (of, the, world, and)     15   8.914403
1330             (be, one, of, the)     11   9.111583
1329              (is, up, to, the)     14   9.336419
1328             (it, is, not, the)     11   9.351034
1327             (to, learn, in, a)     10   9.422595
1326             (a, part, of, the)     10   9.478018
1325             (to, a, sense, of)     10   9.519370
1324        (important, it, is, to)     10   9.618614
1323           (to, working, in, a)     14   9.619860
1322             (it, can, have, a)     16   9.807884
1321     (both, the, students, and)     10   9.892580
1320           (to, the, power, of)     11  10.048092
1319      (in, the, classroom, and)     13  10.061602
1318        (to, have, a, positive)     11  10.146131
1317        (and, the, ability, to)     27  10.153844

In [ ]:
# less than 11
lowish_pmi_quad = df_quadgram.loc[df_quadgram['PMI'] < 11]

In [ ]:
len(lowish_pmi_quad)

43

In [ ]:
len(df_quadgram)

1337

In [ ]:
lowish_pmi_quad['Count'].mean()

16.13953488372093

In [ ]:
low_pmi_df.sort_values("Count", ascending=False)

Count       PMI            Phrases
116      88  4.435353        to  have  a
150      77  4.923616          it  is  a
186      70  4.110780        it  is  the
218      61  3.991271        and  it  is
255      57  3.701502    to  the  school
...     ...       ...                ...
3783     10  0.845095             and  a
3784     10  4.579665    of  failure  in
3785     10  2.516364            on  the
3786     10  4.237282  as  students  can
4327     10 -2.093969       the  to  the

[332 rows x 3 columns]

In [ ]:
low_pmi_df.to_csv('low pmi trigrams.csv',index=False)

In [ ]:
low_pmi_df2

Phrase  Count       PMI
1333             (and, it, can, be)     19  8.804104
1322             (it, can, have, a)     16  9.807884
1331          (of, the, world, and)     15  8.914403
1329              (is, up, to, the)     14  9.336419
1323           (to, working, in, a)     14  9.619860
1336      (the, students, and, the)     13  5.476013
1335    (and, the, opportunity, to)     12  8.164491
1332  (and, understanding, of, the)     12  8.886959
1334            (and, out, of, the)     11  8.614202
1330             (be, one, of, the)     11  9.111583
1328             (it, is, not, the)     11  9.351034
1325             (to, a, sense, of)     10  9.519370
1326             (a, part, of, the)     10  9.478018
1327             (to, learn, in, a)     10  9.422595
1324        (important, it, is, to)     10  9.618614
1321     (both, the, students, and)     10  9.892580

In [ ]:
# find average count of low pmi
low_pmi_df['Count'].mean()

16.768072289156628

In [ ]:
from nltk import everygrams

list(everygrams(one_string1.split(), 1, 5))

In [ ]:
# 1 find everygrams 2) find their pmis and counts 3) put in dataframe


# Method 1 : Doesnt include everything 
# concat bi,tir and quadgrams into one dataframe

df_bigram = pd.read_csv('bigram phrases.csv')

In [ ]:
df_bigram.drop(columns=['Unnamed: 0'], inplace=True)

In [ ]:
df_bigram

Phrase  Count       PMI
0              ('it', 'is')   2400  4.606781
1             ('of', 'the')   2285  1.802598
2             ('can', 'be')   2258  4.634889
3             ('in', 'the')   2215  2.145940
4             ('on', 'the')   1456  2.924906
...                     ...    ...       ...
8629     ('of', 'teaching')     10  2.256682
8630      ('of', 'reality')     10  2.274829
8631    ('have', 'several')     10  2.275628
8632  ('optimistic', 'and')     10  2.276579
8633     ('and', 'falling')     10  2.960105

[8634 rows x 3 columns]

In [ ]:
print(df_bigram['Phrase'].dtype)


object


In [ ]:
phrases = [' '.join(eval(phrase)) for phrase in df_bigram['Phrase']]


In [ ]:
phrases

In [ ]:
df_bigram['Phrase'] = phrases

In [ ]:
df_bigram

Phrase  Count       PMI
0              it is   2400  4.606781
1             of the   2285  1.802598
2             can be   2258  4.634889
3             in the   2215  2.145940
4             on the   1456  2.924906
...              ...    ...       ...
8629     of teaching     10  2.256682
8630      of reality     10  2.274829
8631    have several     10  2.275628
8632  optimistic and     10  2.276579
8633     and falling     10  2.960105

[8634 rows x 3 columns]

In [ ]:
df_trigram

Count        PMI                Phrases
0       788  11.326107      it  is  important
1       694   9.355931      is  important  to
2       512  10.086778         on  the  other
3       512  10.131633          in  order  to
4       496  12.661344       the  other  hand
...     ...        ...                    ...
4323     10   8.257245    attitude  can  make
4324     10   8.249711         it  harder  to
4325     10   8.228961         will  need  to
4326     10   8.196099  both  beneficial  and
4327     10  -2.093969           the  to  the

[4328 rows x 3 columns]

In [ ]:
# reorder the columns in trigram
# creating the DataFrame
df_data = {'Phrase': df_trigram['Phrases'], 
           'Count': df_trigram['Count'], 
           'PMI': df_trigram['PMI']}
df_triram2 = pd.DataFrame(data = df_data)

In [ ]:
df_trigram2 = df_triram2
df_trigram2

Phrase  Count        PMI
0         it  is  important    788  11.326107
1         is  important  to    694   9.355931
2            on  the  other    512  10.086778
3             in  order  to    512  10.131633
4          the  other  hand    496  12.661344
...                     ...    ...        ...
4323    attitude  can  make     10   8.257245
4324         it  harder  to     10   8.249711
4325         will  need  to     10   8.228961
4326  both  beneficial  and     10   8.196099
4327           the  to  the     10  -2.093969

[4328 rows x 3 columns]

In [ ]:
df_quadgram

Phrase  Count        PMI
1335         and the opportunity to     12   8.164491
1334                 and out of the     11   8.614202
1333                  and it can be     19   8.804104
1332       and understanding of the     12   8.886959
1331               of the world and     15   8.914403
...                             ...    ...        ...
4        after school homework club     34  27.850734
3          and early 20th centuries     11  27.860815
2               setting our aim too     19  27.918236
1     seeking guidance from experts     18  28.166498
0                   our aim too low     10  29.144632

[1336 rows x 3 columns]

In [ ]:
merged_df = pd.concat([df_quadgram,df_trigram2,df_bigram])

In [ ]:
merged_df

Phrase  Count       PMI
1335    and the opportunity to     12  8.164491
1334            and out of the     11  8.614202
1333             and it can be     19  8.804104
1332  and understanding of the     12  8.886959
1331          of the world and     15  8.914403
...                        ...    ...       ...
8629               of teaching     10  2.256682
8630                of reality     10  2.274829
8631              have several     10  2.275628
8632            optimistic and     10  2.276579
8633               and falling     10  2.960105

[14298 rows x 3 columns]

In [ ]:
merged_df = merged_df.sort_values('PMI', ascending =False)

In [ ]:
merged_df.head(15)

Phrase  Count        PMI
0                  our aim too low     10  29.144632
1    seeking guidance from experts     18  28.166498
2              setting our aim too     19  27.918236
3         and early 20th centuries     11  27.860815
4       after school homework club     34  27.850734
5        already mastered you will     14  27.840469
6   an increasingly popular option     12  27.840215
7           wonderful how much may     13  27.519006
8          mastered you will never     13  27.193504
9             here 's a simplified     10  27.150397
10     after school homework clubs     10  27.062831
11          small acts of kindness     24  27.018153
12             no matter how small     12  26.839840
13            four day school week     91  26.817121
14        do something beyond what     18  26.624140

In [ ]:
merged_df = merged_df.drop(axis=0 , index =[3,4])

In [ ]:
merged_df

Phrase  Count        PMI
0                   our aim too low     10  29.144632
1     seeking guidance from experts     18  28.166498
2               setting our aim too     19  27.918236
3          and early 20th centuries     11  27.860815
4        after school homework club     34  27.850734
...                             ...    ...        ...
7784                        can and     10  -5.166051
3636                        the the     22  -5.549096
5776                          of to     14  -5.558930
6690                         and of     12  -5.691381
4160                         the to     19  -5.771505

[14298 rows x 3 columns]

In [ ]:
merged_df.to_csv('merged phrases.csv')

In [ ]:
df_bigram2

0        
1        
2        
3        
4        
       ..
8629     
8630     
8631     
8632     
8633     
Name: Phrase, Length: 8634, dtype: object

In [ ]:
# remove parentheses 
clean_phrase = []
for phrases in df_bigram['Phrase']:
    no_brack = ' '.join(phrases)
    clean_phrase.append(no_brack)

# Analysis on chat-gpt with Q for prop. nouns and periods 

In [ ]:
# upload the reddit words with Q ( worked on in GroD Dataset.ipynb)
# reddit_q = pd.read_csv('Q Reddit GPT Words.csv')

>>>> Reddit GPT Q and T 

In [42]:
df =pd.read_csv(r'C:\Users\Admin\HC3 EDA\reddit_filtered_dataset.csv')
gpt_df = df[(df['Labels']==1)]

In [2]:
import pandas as pd

In [43]:
import nltk
from nltk import *

def preprocess(sent):
    sent = nltk.word_tokenize(sent)
    sent = nltk.pos_tag(sent)
    return sent

clean_gpt = []
try:
  for sentence in gpt_df['Data']:

    # process sentences
    cleane = preprocess(sentence)
    clean_gpt.append(cleane)

except KeyError:
  print("Column 'Data' not found")

In [44]:
QT_reddit_gpt =[]
for pair in clean_gpt:

    for word, tag in pair:
        if tag != "NNP" and tag != ":" and tag != "''" and tag != "``" and tag != ".":
            QT_reddit_gpt.append(word)

        elif tag == "NNP": 
            q = "Q"
            QT_reddit_gpt.append(q)

        elif tag == ":":
            q = "T"   # T for punct
            QT_reddit_gpt.append(q)

        elif tag == ".":
            q = "T"
            QT_reddit_gpt.append(q)
            
clean_QT_reddit =[]
for word in QT_reddit_gpt:
    if word != "Q" and word != "T":

        lowe= word.lower()
        clean_QT_reddit.append(lowe)
    elif word == "Q":
        q ="Q"
        clean_QT_reddit.append(q)
    elif word == "T":
        q = "T"
        clean_QT_reddit.append(q)

In [46]:
# save to df
tem_df_1 = pd.DataFrame({'Word':clean_QT_reddit})
tem_df_1.to_csv('Q T Reddit GPT Words.csv')

In [135]:
tem_df_1.value_counts().head(10)

Word
the     6160
T       5576
,       5550
Q       3705
of      3106
and     3013
to      2819
a       2493
in      1909
is      1658
Name: count, dtype: int64

In [2]:
# HC3 datasets 
import json

with open(r"C:\Users\Admin\HC3 json file") as f:
  data = json.load(f)

# parse the gpt 
gpt_answers = []

for row in data['rows']:
  gpt = row['row']['chatgpt_answers']
  gpt_answers.append(gpt)




In [167]:
gpt_answers

[['There are many different best seller lists that are published by various organizations, and the New York Times is just one of them. The New York Times best seller list is a weekly list that ranks the best-selling books in the United States based on sales data from a number of different retailers. The list is published in the New York Times newspaper and is widely considered to be one of the most influential best seller lists in the book industry. \nIt\'s important to note that the New York Times best seller list is not the only best seller list out there, and there are many other lists that rank the top-selling books in different categories or in different countries. So it\'s possible that a book could be a best seller on one list but not on another. \nAdditionally, the term "best seller" is often used more broadly to refer to any book that is selling well, regardless of whether it is on a specific best seller list or not. So it\'s possible that you may hear about a book being a "be

In [3]:
flat_gpt = [item for sublist in gpt_answers for item in sublist]
flat_gpt

['There are many different best seller lists that are published by various organizations, and the New York Times is just one of them. The New York Times best seller list is a weekly list that ranks the best-selling books in the United States based on sales data from a number of different retailers. The list is published in the New York Times newspaper and is widely considered to be one of the most influential best seller lists in the book industry. \nIt\'s important to note that the New York Times best seller list is not the only best seller list out there, and there are many other lists that rank the top-selling books in different categories or in different countries. So it\'s possible that a book could be a best seller on one list but not on another. \nAdditionally, the term "best seller" is often used more broadly to refer to any book that is selling well, regardless of whether it is on a specific best seller list or not. So it\'s possible that you may hear about a book being a "bes

In [5]:
from collections import Counter
my_counter = Counter()
for lst in flat_gpt:
    indiv =  word_tokenize(lst)
    my_counter.update(indiv)
# for word in Words:
#     my_counter.update(word)

In [18]:
for lst in flat_gpt:
    print(lst)

There are many different best seller lists that are published by various organizations, and the New York Times is just one of them. The New York Times best seller list is a weekly list that ranks the best-selling books in the United States based on sales data from a number of different retailers. The list is published in the New York Times newspaper and is widely considered to be one of the most influential best seller lists in the book industry. 
It's important to note that the New York Times best seller list is not the only best seller list out there, and there are many other lists that rank the top-selling books in different categories or in different countries. So it's possible that a book could be a best seller on one list but not on another. 
Additionally, the term "best seller" is often used more broadly to refer to any book that is selling well, regardless of whether it is on a specific best seller list or not. So it's possible that you may hear about a book being a "best selle

In [15]:
my_counter

Counter({',': 968,
         'the': 950,
         '.': 875,
         'to': 759,
         'and': 661,
         'a': 547,
         'of': 494,
         'is': 412,
         'that': 315,
         'in': 287,
         'it': 266,
         'or': 219,
         'are': 213,
         'for': 213,
         'can': 205,
         'be': 198,
         'have': 137,
         'as': 136,
         'on': 135,
         'they': 135,
         'not': 126,
         'their': 116,
         'you': 115,
         'may': 108,
         "'s": 104,
         'more': 96,
         'by': 91,
         'also': 90,
         'people': 84,
         'It': 83,
         'The': 82,
         'which': 82,
         'with': 80,
         'because': 75,
         'different': 74,
         'other': 74,
         'This': 72,
         'your': 70,
         'important': 69,
         'use': 68,
         'might': 62,
         'from': 60,
         ':': 58,
         'if': 57,
         'has': 57,
         'used': 56,
         'way': 53,
         'such': 52

In [16]:
total_word_count = sum(my_counter.values())
print(total_word_count)


21361


In [177]:
# length of text in words  
cnt = 0
for str in flat_gpt:
   cnt += len(str)
cnt

109877

>> Unresolved discrepancy in two word count methods 

In [25]:
# dbl check THIS METHOD DOESNT WORK
from nltk.tokenize import word_tokenize

tokenized_sents = [word_tokenize(i) for i in flat_gpt]

words_list = []
for tokenized_sent in tokenized_sents:
    words_list.extend(tokenized_sent)

print(len(words_list))  # This should give you the total word count


21361


In [21]:
 #now add Q wherever there is a proper noun , T where there is a period to evulate sentence length 
import nltk
from nltk import *

def preprocess(sent):
    sent = nltk.word_tokenize(sent)
    sent = nltk.pos_tag(sent)
    return sent

clean_gpt = []
try:
  for word in flat_gpt:

    # process sentences
    cleane = preprocess(word)
    clean_gpt.append(cleane)

except KeyError:
  print("Column 'Data' not found")

In [22]:
clean_gpt

[[('There', 'EX'),
  ('are', 'VBP'),
  ('many', 'JJ'),
  ('different', 'JJ'),
  ('best', 'JJS'),
  ('seller', 'NN'),
  ('lists', 'NNS'),
  ('that', 'WDT'),
  ('are', 'VBP'),
  ('published', 'VBN'),
  ('by', 'IN'),
  ('various', 'JJ'),
  ('organizations', 'NNS'),
  (',', ','),
  ('and', 'CC'),
  ('the', 'DT'),
  ('New', 'NNP'),
  ('York', 'NNP'),
  ('Times', 'NNP'),
  ('is', 'VBZ'),
  ('just', 'RB'),
  ('one', 'CD'),
  ('of', 'IN'),
  ('them', 'PRP'),
  ('.', '.'),
  ('The', 'DT'),
  ('New', 'NNP'),
  ('York', 'NNP'),
  ('Times', 'NNP'),
  ('best', 'JJS'),
  ('seller', 'NN'),
  ('list', 'NN'),
  ('is', 'VBZ'),
  ('a', 'DT'),
  ('weekly', 'JJ'),
  ('list', 'NN'),
  ('that', 'WDT'),
  ('ranks', 'VBZ'),
  ('the', 'DT'),
  ('best-selling', 'JJ'),
  ('books', 'NNS'),
  ('in', 'IN'),
  ('the', 'DT'),
  ('United', 'NNP'),
  ('States', 'NNPS'),
  ('based', 'VBN'),
  ('on', 'IN'),
  ('sales', 'NNS'),
  ('data', 'NNS'),
  ('from', 'IN'),
  ('a', 'DT'),
  ('number', 'NN'),
  ('of', 'IN'),
  ('diff

In [23]:
Q_HC3_gpt =[]
for pair in clean_gpt:

    for word, tag in pair:
        if tag != "NNP" and tag != ":" and tag != "''" and tag != "``" and tag != ".":
            Q_HC3_gpt.append(word)

        elif tag == "NNP": 
            q = "Q"
            Q_HC3_gpt.append(q)

        elif tag == ":":
            q = "T"   # T for punct
            Q_HC3_gpt.append(q)

        elif tag == ".":
            q = "T" # T for punctuation 
            Q_HC3_gpt.append(q)
            
clean_HC3_gpt =[]
for word in Q_HC3_gpt:
    if word != "Q" and word != "T":

        lowe= word.lower()
        clean_HC3_gpt.append(lowe)
    elif word == "Q":
        q ="Q"
        clean_HC3_gpt.append(q)

    elif word == "T":
        q ="T"
        clean_HC3_gpt.append(q)

In [28]:
import pandas as pd
clean_HC3_gpt_df =pd.DataFrame({'word':clean_HC3_gpt})
clean_HC3_gpt_df.to_csv('Q T HC3 GPT words.csv', index =False)

In [29]:
len(clean_HC3_gpt_df)

21271

In [ ]:
clean_HC3_gpt_df.value_counts().head(10)

word
the     1032
,        968
T        938
to       764
and      661
a        562
of       494
Q        463
is       412
it       349
Name: count, dtype: int64

>>>> ESSAYS Q and T 

In [32]:
# 
df = pd.read_csv(r"C:\Users\Admin\student vs gpt essays.csv")
gpt_essay = df['source_text']
gpt_essay

0       \nWhen considering the pros and cons of attend...
1       \nOne of the primary arguments for implementin...
2       \nBefore making any decisions about getting in...
3       \nRalph Waldo Emerson once said, "Go confident...
4       \nHuman character traits are shaped by a wide ...
                              ...                        
2416    \nBecoming a surgeon requires a great deal of ...
2417    \nSchools should offer an after school homewor...
2418    \nIt’s human nature to be afraid to make mista...
2419    \nOne of the main debates of 2020 for many stu...
2420    \nThe implementation of a curfew for teenagers...
Name: source_text, Length: 2421, dtype: object

In [33]:
# # char. count
# # for sent in gpt_essay:
# #     for char in sent:
# num_char =[ char for char in sent for sent  in gpt_essay]

# # word count
# # import re
# # word_count = len(re.split(r"\s+", gpt_essay))

# # print(word_count)>>> 
# from collections import Counter
# Counter(num_char)
import re

string_list =[]
for par in gpt_essay:
 ls_ess = str(par)
 words = re.split(r"\s+", ls_ess)
 string_list.append(words)



#gpt_essay = """This is a sample essay generated by GPT-3. It contains several paragraphs discussing different topics in a coherent manner. Overall, the essay flows well and demonstrates a basic understanding of writing structure and composition."""

# word_count = len(words)

# print(word_count)

In [34]:
# word = [ word for word in lst for lst in string_list]
# word

['H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H',
 'H'

In [35]:
# remove \newline
sauber_essay =[]
for sent in gpt_essay:
    sauber_essay.append(sent.replace("\n",""))


In [36]:
clean_gpt_ess = []
try:
  for word in sauber_essay:

    # process sentences
    cleane = preprocess(word)
    clean_gpt_ess.append(cleane)

except KeyError:
  print("Column 'Data' not found")

In [38]:
QT_essay_gpt =[]
for pair in clean_gpt_ess:

    for word, tag in pair:
        if tag != "NNP" and tag != ":" and tag != "''" and tag != "``" and tag != ".":
            QT_essay_gpt.append(word)

        elif tag == "NNP": 
            q = "Q"
            QT_essay_gpt.append(q)

        elif tag == ":":
            q = "T"   # T for punct
            QT_essay_gpt.append(q)

        elif tag == ".":
            q = "T" # T for punctuation 
            QT_essay_gpt.append(q)
            
clean_essay_gpt =[]
for word in QT_essay_gpt:
    if word != "Q" and word != "T":

        lowe= word.lower()
        clean_essay_gpt.append(lowe)
    elif word == "Q":
        q ="Q"
        clean_essay_gpt.append(q)

    elif word == "T":
        q ="T"
        clean_essay_gpt.append(q)

In [ ]:
# matches = [x for x in QT_essay_gpt if x == "Q"]
# # there ARE proper nouns in the QT _essay_gpt list
# matches2 = [x for x in clean_essay_gpt if x == "Q"]
# # clean_+essay_gpt HAS prop nouns
# nnps =[] 
# for item in clean_gpt_ess:
#     for word,label in item:

    
#         if label == 'NNP':
#             nnps.append(word)

In [40]:
tem_df = pd.DataFrame({'word':clean_essay_gpt})
tem_df.to_csv('Q T Essay gpt.csv')

In [41]:
len(tem_df)

539738

>> OBSERVATIONS

In [49]:
clean_HC3_gpt_df.value_counts()

word        
the             1032
,                968
T                938
to               764
and              661
                ... 
incentives         1
incandescent       1
implemented        1
impartial          1
zero               1
Name: count, Length: 2811, dtype: int64

> find number  of proper nouns in each of the following: essay = 5012 , HC3 = 463, and reddit =  3705 
>> count original number of characters in each uncleaned corpus | HC3 = 21271  | essay = 539738 | reddit = 110116
>>> count number of T's = periods  | 21931 in essays |  938 HC3 |  reddit 5576  
>>>> adjectives | 36457 in essays | hc3 1208 | reddit 7151

In [69]:
# average number of words per sentence
# ( total_chars - T) / total_periods

# first find total number of T's in ech dataset
num_ess_Ts = tem_df.value_counts().get('T', 0)

num_red_Ts = tem_df_1.value_counts().get('T',0)

num_HC3_Ts = clean_HC3_gpt_df.value_counts().get('T',0)

# total number of characters in original text , including punctuation
total_chars = len(tem_df) + len(tem_df_1) + len(clean_HC3_gpt_df)

# get the total of words exclude punct
total_periods = 938 + 21931 + 5576
sum_of_words = total_chars - total_periods

# average ; how many words per period ? i.e. what is the average sentence length for the aggreagated texts
average_gpt_sent_length = sum_of_words / total_periods

In [63]:
from mean_sentence_length_function import average_sent_length
# # average sent length for essays
# essays_char_sum = len(tem_df)
# essays_tot_words = essays_char_sum - num_ess_Ts
essays_average = average_sent_length(tem_df,num_ess_Ts)
print(essays_average)
# # aver. sent length for reddit
# red_char_sum = len(tem_df_1)
# red_tot_words = red_char_sum - num_red_Ts
reddit_average = average_sent_length(tem_df_1,num_red_Ts)
print(reddit_average)
# #average HC3 
# HC_char_sum = len(clean_HC3_gpt_df)
# HC_tot_words = HC_char_sum - num_HC3_Ts
HC_average = average_sent_length(clean_HC3_gpt_df,num_HC3_Ts)
print(HC_average)
# add up averges and divide by 3
(HC_average+essays_average+reddit_average)/3

23.610733664675575
18.748206599713058
21.676972281449892


21.345304181946176

>> average of all three averages is 21.345304181946176 . 

In [60]:
def average_sent_length(df,num_periods):
    """
    df = takes of dataframe with one column containing all the words in the corpus , 
    num_periods is a the number of periods in the corpurs.
    -> average number of words per sentence
    assumes df includes periods
    """
    char_sum = len(df)
    tot_words = char_sum - num_periods
    average = tot_words/num_periods
    return average

In [62]:
average_sent_length(tem_df,num_ess_Ts)

23.610733664675575

In [ ]:
tem_df['word'].value_counts().head(30)

word
,           26317
T           21931
to          21518
and         20016
the         17722
a           12609
of          12313
in           9539
can          8276
is           6130
for          5899
it           5710
that         5596
be           5377
Q            5012
students     4404
with         4287
their        4184
as           3856
on           3771
are          3134
have         3068
an           3021
more         2659
i            2634
our          2575
school       2487
or           2404
from         2384
this         2200
Name: count, dtype: int64

In [47]:
len(tem_df_1)

110116

In [50]:
tem_df_1.value_counts().head(20)

Word
the     6160
T       5576
,       5550
Q       3705
of      3106
and     3013
to      2819
a       2493
in      1909
is      1658
it      1186
that    1129
as       805
or       791
are      786
can      730
for      730
have     602
on       579
's       576
Name: count, dtype: int64

In [ ]:
# prep POS tagged lists

#essay
labels_list_ess =[]
for word in tem_df['word']:
    pari =preprocess(word)
    labels_list_ess.append(pari)


In [88]:

# reddit
labels_list_red =[]
for word in tem_df_1['Word']:
    pari =preprocess(word)
    labels_list_red.append(pari)
    

In [89]:
#hc3
labels_list_hc3 =[]
for word in clean_HC3_gpt_df['word']:
    pair = preprocess(word)
    labels_list_hc3.append(pair)

In [84]:
# count num of adj
adjectives_ess = []
for lst in labels_list_ess:
    for tup in lst:
        #word, tag = tup
        word ,tag = tup 
        if tag == "JJ":
            adjectives_ess.append(word)

       
        

['physical',
 'potential',
 'own',
 'other',
 'social',
 'potential',
 'technical',
 'home-based',
 'self-disciplined',
 'full',
 'own',
 'physical',
 'home-based',
 'effective',
 'individual',
 'available',
 'four-day',
 'potential',
 'many',
 'four-day',
 'several',
 'potential',
 'extra',
 'four-day',
 'difficult',
 'certain',
 'such',
 'four-day',
 'important',
 'potential',
 'potential',
 'negative',
 'such',
 'important',
 'clear',
 'much',
 'many',
 'poor',
 'comfortable',
 'current',
 'live',
 'valuable',
 'current',
 'potential',
 'many',
 'educational',
 'extra',
 'hard',
 'difficult',
 'short',
 'strong',
 'ethic',
 'own',
 'great',
 'invaluable',
 'full',
 'potential',
 'wide',
 'major',
 'moral',
 'positive',
 'powerful',
 'significant',
 'academic',
 'potential',
 'different',
 'intellectual',
 'social',
 'cultural',
 'advantageous',
 'popular',
 'social',
 'several',
 'accessible',
 'other',
 'several',
 'difficult',
 'regular',
 'personal',
 'several',
 'successful',
 '

In [92]:
# count num of adj red
adjectives_red = []
for lst in labels_list_red:
    for tup in lst:
        #word, tag = tup
        word ,tag = tup 
        if tag == "JJ":
            adjectives_red.append(word)


In [94]:
# count num of adj
adjectives_hc = []
for lst in labels_list_hc3:
    for tup in lst:
        #word, tag = tup
        word ,tag = tup 
        if tag == "JJ":
            adjectives_hc.append(word)


In [96]:
len(adjectives_red)

7151

>>>>>>>>>>>>>compare to humans

In [ ]:
import pandas as pd

In [ ]:
# df_reddit =pd.read_csv(r'C:\Users\Admin\HC3 EDA\reddit_filtered_dataset.csv')

In [ ]:
# df_reddit_human = df_reddit[df_reddit['Labels']==0]

In [ ]:
# df_reddit_human.drop(columns=['Labels'],inplace=True)

In [ ]:
# import nltk
# from nltk import *

In [ ]:
# # clean reddit human
# def preprocess(sent):
#     sent = nltk.word_tokenize(sent)
#     sent = nltk.pos_tag(sent)
#     return sent

# clean1 = []


# try:
#   for sentence in df_reddit_human["Data"]:

#     # process sentences
#     cleane = preprocess(sentence)
#     clean1.append(cleane)

# except KeyError:
#   print("Column 'Data' not found")

In [ ]:
# # reomve PRP$, ':'and "''" replace w q . ALso  '``' and "''" 
# good_words =[]
# for pair in clean1:

#     for word, tag in pair:
#         if tag != "NNP" and tag != ":" and tag != "''" and tag != "``" and tag != ".":
#             good_words.append(word)

#         elif tag == "NNP": 
#             q = "Q"
#             good_words.append(q)

#         elif tag == ":":
#             q = "Q" 
#             good_words.append(q)

        # elif tag == ".":
        #     q = "Q"
        #     good_words.append(q)

        # elif tag == "NNP" and tag == ":" and tag == "." :
        #     q = "Q"
        # # replace word with q
        #     good_words.append(q)

    #elif tag == '``' and tag == "''":
        # just remove dont replace with anything


In [ ]:
# find bi tri and quad grams skipping over Q 


In [ ]:
# df_essay = pd.read_csv(r"C:\Users\Admin\student vs gpt essays.csv")

In [ ]:
# human_essay = df_essay['text']

In [ ]:
# clean essay 

In [ ]:
# # HC3 datasets 
# import json

# with open(r"C:\Users\Admin\HC3 json file") as f:
#   data = json.load(f)

# # parse the human
# human_answers = []

# for row in data['rows']:
#   human = row['row']['human_answers']
#   human_answers.append(human)

# # flatten answers
# flat_human = [item for sublist in human_answers for item in sublist]
# flat_human 